File: python/flink_agents/plan/tools/bash/bash_validator.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
################################################################################
"""AST-based bash command validation using tree-sitter-bash.

Parses the command with tree-sitter-bash and walks the AST. Any named node
whose type is not in the allowed set (e.g. ``command_substitution``,
``process_substitution``, ``subshell``, ``for_statement``) causes the whole
command to be rejected. Every ``command`` node's name is checked against the
``allowed_commands`` allowlist or resolved under ``allowed_script_dirs``.

This lets the tool accept natural shell constructs like pipes, ``&&`` / ``||``
chains and file-descriptor-only redirections while blocking common injection
vectors (``$()``, backticks, file redirects, execution-changing environment
assignments, control flow, etc.).
"""

from __future__ import annotations

from functools import lru_cache
from pathlib import Path
from typing import TYPE_CHECKING, List

if TYPE_CHECKING:
    from tree_sitter import Node, Parser


# Named AST node types we accept. Anything named but missing is treated as a
# potentially dangerous shell construct and rejected. Unnamed nodes (literal
# punctuation like ``|``, ``&&``, ``(``, ``$(``) are always allowed — they're
# just syntax tokens, not semantic structures.
_ALLOWED_NAMED = frozenset(
    {
        "program",
        "command",
        "command_name",
        "pipeline",
        "list",
        "redirected_statement",
        "file_redirect",
        "file_descriptor",
        "variable_assignment",
        "variable_name",
        "special_variable_name",  # $@ $? $* $#
        "word",
        "string",
        "string_content",
        "raw_string",
        "ansi_c_string",
        "translated_string",
        "concatenation",
        "number",
        "simple_expansion",  # $VAR
        "expansion",  # ${VAR}
        "binary_expression",
        "unary_expression",
        "parenthesized_expression",
        "array",
    }
)

_BLOCKED_ENVIRONMENT_VARIABLES = frozenset(
    {"PATH", "BASH_ENV", "ENV", "SHELLOPTS", "CDPATH"}
)
_DYNAMIC_LOADER_VARIABLE_PREFIXES = ("LD_", "DYLD_")
_FD_REDIRECT_OPERATORS = frozenset({"<&", ">&"})
_FD_CLOSE_OPERATORS = frozenset({"<&-", ">&-"})


@lru_cache(maxsize=1)
def _get_parser() -> Parser:
    """Return a cached tree-sitter parser configured with the bash grammar."""
    import tree_sitter_bash
    from tree_sitter import Language, Parser

    return Parser(Language(tree_sitter_bash.language()))


def validate_command(
    command: str,
    allowed_commands: List[str],
    allowed_script_dirs: List[str],
    cwd: str | None = None,
) -> str | None:
    """Validate a bash command.

    Returns ``None`` if the command is allowed, or an error string otherwise.
    """
    if not command.strip():
        return "Empty command."

    try:
        tree = _get_parser().parse(command.encode("utf-8"))
    except Exception as exc:
        return f"Failed to parse command: {exc}"

    root = tree.root_node
    if root.has_error:
        return "Command has syntax errors."
    if not root.children:
        return "Empty command."

    return _walk(root, allowed_commands, allowed_script_dirs, cwd)


def _walk(
    node: Node,
    allowed_commands: List[str],
    allowed_script_dirs: List[str],
    cwd: str | None,
) -> str | None:
    if node.is_named and node.type not in _ALLOWED_NAMED:
        snippet = node.text.decode("utf-8", errors="replace")[:80]
        return f"Disallowed shell construct '{node.type}' in: {snippet!r}"
    if node.type == "variable_assignment" and (
        node.parent is None or node.parent.type != "command"
    ):
        return "Standalone variable assignment without an executable is not allowed."
    if node.type == "file_redirect" and not _is_fd_only_redirect(node):
        return (
            "File redirects are not allowed; only file-descriptor duplication "
            "and closure are permitted."
        )
    if node.type == "variable_assignment":
        name_node = node.child_by_field_name("name")
        if name_node is not None:
            name = name_node.text.decode("utf-8", errors="replace")
            if _is_blocked_environment_variable(name):
                return f"Environment variable assignment '{name}' is not allowed."
    if node.type == "command":
        err = _validate_command_node(node, allowed_commands, allowed_script_dirs, cwd)
        if err is not None:
            return err
    for child in node.children:
        err = _walk(child, allowed_commands, allowed_script_dirs, cwd)
        if err is not None:
            return err
    return None


def _is_fd_only_redirect(node: Node) -> bool:
    """Return whether a redirect only duplicates or closes a file descriptor."""
    operator = next((child.type for child in node.children if not child.is_named), None)
    destination = node.child_by_field_name("destination")
    if operator in _FD_CLOSE_OPERATORS:
        return destination is None
    return (
        operator in _FD_REDIRECT_OPERATORS
        and destination is not None
        and (
            destination.type == "number"
            or (
                destination.type == "word"
                and destination.text.endswith(b"-")
                and destination.text[:-1].isdigit()
            )
        )
    )


def _is_blocked_environment_variable(name: str) -> bool:
    return name in _BLOCKED_ENVIRONMENT_VARIABLES or name.startswith(
        _DYNAMIC_LOADER_VARIABLE_PREFIXES
    )


def _validate_command_node(
    node: Node,
    allowed_commands: List[str],
    allowed_script_dirs: List[str],
    cwd: str | None,
) -> str | None:
    name_node = node.child_by_field_name("name")
    if name_node is None:
        # Fail closed for constructs such as bare variable assignments. Bash
        # can later reinterpret their values in arithmetic contexts.
        return "Command without an executable is not allowed."
    executable = name_node.text.decode("utf-8", errors="replace")
    if executable in allowed_commands:
        return None
    if is_under_allowed_dirs(executable, allowed_script_dirs, cwd):
        return None
    return (
        f"Command '{executable}' is not allowed. "
        f"Allowed commands: {sorted(allowed_commands)}. "
        f"Allowed script dirs: {sorted(allowed_script_dirs)}."
    )


def is_under_allowed_dirs(
    path_str: str,
    allowed_dirs: List[str],
    cwd: str | None = None,
) -> bool:
    """Return True if ``path_str`` resolves to a path under any allowed dir.

    When ``cwd`` is given, relative ``path_str`` is resolved against ``cwd``.
    """
    try:
        base = Path(path_str)
        if not base.is_absolute() and cwd is not None:
            base = Path(cwd) / base
        path = base.resolve()
    except (OSError, ValueError):
        return False
    for allowed in allowed_dirs:
        try:
            allowed_root = Path(allowed).resolve()
        except (OSError, ValueError):
            continue
        try:
            path.relative_to(allowed_root)
        except ValueError:
            continue
        else:
            return True
    return False

File: plan/src/main/java/org/apache/flink/agents/plan/tools/bash/BashValidator.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.agents.plan.tools.bash;

import org.treesitter.TSNode;
import org.treesitter.TSParser;
import org.treesitter.TSTree;
import org.treesitter.TreeSitterBash;

import javax.annotation.Nullable;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.Set;

/**
 * AST-based bash command validator backed by tree-sitter-bash.
 *
 * <p>Mirrors the Python {@code flink_agents.plan.tools.bash.bash_validator}: walks the parsed AST
 * and rejects any named node whose type is not on the {@link #ALLOWED_NAMED} allowlist (e.g. {@code
 * command_substitution}, {@code subshell}, {@code for_statement}, etc.); for every {@code command}
 * node it requires the executable to be either in {@code allowedCommands} or to resolve to a path
 * under one of {@code allowedScriptDirs}.
 */
public final class BashValidator {

    /**
     * Named AST node types we accept. Anything named but missing from this set is treated as a
     * potentially dangerous shell construct and rejected. Unnamed nodes (literal punctuation like
     * {@code |}, {@code &&}, {@code (}) are always allowed — they're just syntax tokens.
     *
     * <p>Kept in sync with the Python {@code _ALLOWED_NAMED} set.
     */
    public static final Set<String> ALLOWED_NAMED =
            Set.of(
                    "program",
                    "command",
                    "command_name",
                    // `export VAR=...`, `readonly`, `declare`, `local`, `typeset`
                    "declaration_command",
                    "pipeline",
                    "list",
                    "redirected_statement",
                    "file_redirect",
                    "file_descriptor",
                    "variable_assignment",
                    "variable_name",
                    "special_variable_name", // $@ $? $* $#
                    "word",
                    "string",
                    "string_content",
                    "raw_string",
                    "ansi_c_string",
                    "translated_string",
                    "concatenation",
                    "number",
                    "simple_expansion", // $VAR
                    "expansion", // ${VAR}
                    "arithmetic_expansion", // $((...))
                    "binary_expression",
                    "unary_expression",
                    "parenthesized_expression",
                    "array");

    private static final Object PARSER_LOCK = new Object();
    private static volatile TSParser parser;

    private BashValidator() {}

    private static TSParser parser() {
        TSParser p = parser;
        if (p == null) {
            synchronized (PARSER_LOCK) {
                p = parser;
                if (p == null) {
                    TSParser created = new TSParser();
                    created.setLanguage(new TreeSitterBash());
                    parser = created;
                    p = created;
                }
            }
        }
        return p;
    }

    /**
     * Validate a bash command. Returns {@link Optional#empty()} when allowed, or a non-empty
     * descriptive error otherwise.
     */
    public static Optional<String> validate(
            String command,
            List<String> allowedCommands,
            List<String> allowedScriptDirs,
            @Nullable String cwd) {
        if (command == null || command.trim().isEmpty()) {
            return Optional.of("Empty command.");
        }
        TSTree tree;
        synchronized (PARSER_LOCK) {
            tree = parser().parseString(null, command);
        }
        TSNode root = tree.getRootNode();
        if (root.hasError()) {
            return Optional.of("Command has syntax errors.");
        }
        if (root.getChildCount() == 0) {
            return Optional.of("Empty command.");
        }
        return walk(root, command, allowedCommands, allowedScriptDirs, cwd);
    }

    private static Optional<String> walk(
            TSNode node,
            String command,
            List<String> allowedCommands,
            List<String> allowedScriptDirs,
            @Nullable String cwd) {
        if (node.isNamed() && !ALLOWED_NAMED.contains(node.getType())) {
            String snippet = nodeText(node, command);
            if (snippet.length() > 80) {
                snippet = snippet.substring(0, 80);
            }
            return Optional.of(
                    "Disallowed shell construct '" + node.getType() + "' in: '" + snippet + "'");
        }
        if ("command".equals(node.getType())) {
            Optional<String> err =
                    validateCommand(node, command, allowedCommands, allowedScriptDirs, cwd);
            if (err.isPresent()) {
                return err;
            }
        }
        for (int i = 0; i < node.getChildCount(); i++) {
            Optional<String> err =
                    walk(node.getChild(i), command, allowedCommands, allowedScriptDirs, cwd);
            if (err.isPresent()) {
                return err;
            }
        }
        return Optional.empty();
    }

    private static Optional<String> validateCommand(
            TSNode commandNode,
            String command,
            List<String> allowedCommands,
            List<String> allowedScriptDirs,
            @Nullable String cwd) {
        TSNode nameNode = commandNode.getChildByFieldName("name");
        if (nameNode == null || nameNode.isNull()) {
            // Bare variable-assignment parsed as command — nothing to validate.
            return Optional.empty();
        }
        String executable = nodeText(nameNode, command);
        if (allowedCommands.contains(executable)) {
            return Optional.empty();
        }
        if (isUnderAllowedDirs(executable, allowedScriptDirs, cwd)) {
            return Optional.empty();
        }
        Set<String> sortedCommands = new HashSet<>(allowedCommands);
        Set<String> sortedDirs = new HashSet<>(allowedScriptDirs);
        return Optional.of(
                "Command '"
                        + executable
                        + "' is not allowed. Allowed commands: "
                        + sortedCommands
                        + ". Allowed script dirs: "
                        + sortedDirs
                        + ".");
    }

    /** Return true when {@code pathStr} resolves to a path under any of the allowed dirs. */
    public static boolean isUnderAllowedDirs(
            String pathStr, List<String> allowedDirs, @Nullable String cwd) {
        Path base;
        try {
            base = Path.of(pathStr);
        } catch (Exception e) {
            return false;
        }
        if (!base.isAbsolute() && cwd != null) {
            base = Path.of(cwd).resolve(base);
        }
        Path resolved;
        try {
            resolved = base.toAbsolutePath().toRealPath();
        } catch (IOException e) {
            try {
                resolved = base.toAbsolutePath().normalize();
            } catch (Exception ee) {
                return false;
            }
        } catch (Exception e) {
            return false;
        }
        for (String allowed : allowedDirs) {
            try {
                Path allowedRoot;
                try {
                    allowedRoot = Path.of(allowed).toAbsolutePath().toRealPath();
                } catch (IOException io) {
                    allowedRoot = Path.of(allowed).toAbsolutePath().normalize();
                }
                if (resolved.startsWith(allowedRoot)) {
                    return true;
                }
            } catch (Exception ignored) {
                // skip invalid allowed root
            }
        }
        return false;
    }

    private static String nodeText(TSNode node, String command) {
        byte[] bytes = command.getBytes(StandardCharsets.UTF_8);
        int start = node.getStartByte();
        int end = node.getEndByte();
        if (start < 0 || end > bytes.length || start > end) {
            return "";
        }
        return new String(bytes, start, end - start, StandardCharsets.UTF_8);
    }
}
